contracts/os/andromeda-kernel/src/execute.rs

```diff
@@ -39,101 +35,104 @@ pub fn send(ctx: ExecuteContext, message: AMPMsg) -> Result<Response, ContractEr
-//         .add_attribute("receiving_kernel_address:{}", channel_info.kernel_address)
-//         .add_attribute("chain:{}", chain)
-//         .add_submessage(SubMsg {
-//             id: ReplyId::TransferFunds.repr(),
-//             msg: CosmosMsg::Ibc(msg),
-//             gas_limit: None,
-//             reply_on: cosmwasm_std::ReplyOn::Always,
-//         }))
-// }
+pub fn transfer_reply(
+    ctx: ExecuteContext,
+    packet_sequence: String,
+) -> Result<Response, ContractError> {
+    //TODO Only the authorized address to handle replies can call this function
+    let ics20_packet_info =
+        CHANNEL_TO_EXECUTE_MSG.load(ctx.deps.storage, packet_sequence.clone())?;
+
+    let chain =
+        ics20_packet_info
+            .recipient
+            .get_chain()
+            .ok_or_else(|| ContractError::InvalidPacket {
+                error: Some("Chain not provided".to_string()),
+            })?;
+
+    let channel_info = CHAIN_TO_CHANNEL
+        .may_load(ctx.deps.storage, chain)?
+        .ok_or_else(|| ContractError::InvalidPacket {
+            error: Some(format!("Channel not found for chain {}", chain)),
+        })?;
+
+    handle_ibc_transfer_funds_reply(
+        ctx.deps,
+        ctx.info,
+        ctx.env,
+        ctx.amp_ctx,
+        0,
+        channel_info,
+        ics20_packet_info,
+    )
+}
+
+fn handle_ibc_transfer_funds_reply(
+    _deps: DepsMut,
+    _info: MessageInfo,
+    env: Env,
+    _ctx: Option<AMPPkt>,
+    sequence: u64,
+    channel_info: ChannelInfo,
+    ics20_packet_info: Ics20PacketInfo,
+) -> Result<Response, ContractError> {
+    ensure!(
+        !Binary::default().eq(&ics20_packet_info.message),
+        ContractError::InvalidPacket {
+            error: Some("The transfer funds reply must contain a message".to_string())
+        }
+    );
+    let ics20_packet_info = ics20_packet_info.clone();
+    let chain =
+        ics20_packet_info
+            .recipient
+            .get_chain()
+            .ok_or_else(|| ContractError::InvalidPacket {
+                error: Some("Chain not provided in recipient".to_string()),
+            })?;
+
+    let channel = channel_info
+        .direct_channel_id
+        .ok_or_else(|| ContractError::InvalidPacket {
+            error: Some(format!("Direct channel not found for chain {}", chain)),
+        })?;
+    #[cfg(target_arch = "wasm32")]
+    let adjusted_funds = Coin::new(
+        ics20_packet_info.funds.amount.u128(),
+        hash_denom_trace(
+            format!(
+                "{}/{}/{}",
+                "transfer", channel, ics20_packet_info.funds.denom
+            )
+            .as_str(),
+        ),
+    );
+
+    // Funds are not correctly hashed when using cw-orchestrator so instead we construct the denom manually
+    #[cfg(not(target_arch = "wasm32"))]
+    let adjusted_funds = Coin::new(
+        ics20_packet_info.funds.amount.u128(),
+        format!("ibc/{}/{}", channel, ics20_packet_info.funds.denom),
```

Comment:
This should be the receiving end's channel ID rather than sending end's channel ID. This should be available in the ICS20 packet info I think?